### PR TITLE
feat(upload): showUploadList options showXxxIcon support function

### PR DESCRIPTION
**Pull request by guoyunhe**
…n and showRemoveIcon support function

[中文版模板 / Chinese template](https://github.com/ant-design/ant-design/blob/master/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)

### 🤔 This is a ...

- [x] New feature
- [ ] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Internationalization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Workflow
- [ ] Other (about what?)

### 💡 Background and solution

有些时候我们可能需要根据文件的具体信息来决定是否要隐藏、展示某些按钮。这个 PR 增加了函数支持，比如下面这个例子，在文件审核通过后，禁止删除：

```jsx
<Upload
  showFileList={{
    showRemoveIcon: file => !file.response.approved
  }}
/>
```

### 📝 Changelog

- Use a developer-oriented tone and narrative style.
- Describe the user's first-hand experience of the issue and its impact on developers, rather than your solution approach.
- Refer to: https://ant.design/changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English | feat(upload): showUploadList showXxxIcon options support function |
| 🇨🇳 Chinese | feat(upload): showUploadList showXxxIcon 选项支持函数 |

### CODE REVIEW

File: components/upload/UploadList/ListItem.tsx [hunk 1]
```diff
@@ -25,9 +25,9 @@ export interface ListItemProps {
   items: UploadFile[];
   listType?: UploadListType;
   isImgUrl?: (file: UploadFile) => boolean;
-  showRemoveIcon?: boolean;
-  showDownloadIcon?: boolean;
-  showPreviewIcon?: boolean;
+  showRemoveIcon?: boolean | ((file: UploadFile) => boolean);
```
Comment (zombieJ) [inline components/upload/UploadList/ListItem.tsx:28]: test case pls

**Comment by li-jia-nan** [comment]
这个应该提交到 feature 分支

**Review by zombieJ — DISMISSED** [review]

**Comment by zombieJ** [comment]
Coverage 掉了

File: components/upload/interface.ts [hunk 2]
```diff
@@ -51,9 +51,9 @@ export interface UploadChangeParam<T = UploadFile> {
 
 export interface ShowUploadListInterface<T = any> {
   extra?: React.ReactNode | ((file: UploadFile<T>) => React.ReactNode);
-  showRemoveIcon?: boolean;
-  showPreviewIcon?: boolean;
-  showDownloadIcon?: boolean;
+  showRemoveIcon?: boolean | ((file: UploadFile<T>, disabled: boolean) => boolean);
```
Comment (zombieJ) [inline components/upload/interface.ts:54]: 这不行，不能直接把单 state 作为 param 传递。以后一个个 state 传就不能维护了。应该是给一个 info 里面包含：

```tsx
((file: UploadFile<T>, info: { disabled: boolean }) => boolean)
```

**Review by zombieJ — APPROVED** [review]

**Pull request merged**